Remove hardwired MaxChemLev; now compute this as the 1 hPa level and store in State_Met

yantosca opened the pull request:
### Name and Institution (Required)

Name: Bob Yantosca
Institution: Harvard + GCST

### Describe the update
This is the companion PR to #284, in which we do the following:

1. Remove `State_Grid%MaxTropLev`, `State_Grid%MaxStratLev`, and `State_Grid%MaxChemLev` fields.

2. Add `State_Met%MaxChemLev` and `State_Met%MaxStratLev` fields

3. Compute `State_Met%MaxStratLev` as the 1 hPa level instead of hardwiring it (for GMAO fields only.  GCAP/ModelE2.1 fields still set these to hardwired values in order to ensure backwards compatibility with previous model versions).

4. Set `State_Met%MaxChemLev` equal to `State_Met%MaxStratLev`, so that chemistry will be done to the top of the stratosphere.  This can be changed in the future to push chemistry higher up in the atmosphere.

5. Reference `State_Met%MaxChemLev` and `State_Met%MaxStratLev` in various routines.  In some cases this has involved adding `State_Met` as an argument to routines.

6. Simplify the logic in `gc_grid_mod.F90` where `isGMAO` and `State_Grid%NativeNZ` are computed.

### Expected changes
This is a no-diff-to-benchmark update.

### Related Github Issue
- Closes #284

yantosca committed:
Now compute MaxChemLev and MaxStratLev as the 1 hPa level

Headers/state_grid_mod.F90
- Removed MaxChemLev, MaxStratLev, MaxTropLev fields from State_Grid

Headers/state_met_mod.F90
- Added MaxChemLev and MaxStratLev to State_Met

GeosUtil/pressure_mod.F90
- Added Init_MaxChemLev routine
- Call Init_MaxChemLev from Init_Pressure
- Pass State_Met argument to Init_Pressure
- Now define State_Met%MaxChemLev and State_Met%MaxStratLev as the
  level where pressure = 1 hPa

GeosCore/calc_met_mod.F90
- Now reference State_Met%MaxStratLev instead of State_Grid%MaxStratLev
- Now reference State_Met%MaxChemLev instead of State_Grid%MaxChemLev

GeosCore/cldj_interface_mod.F90
- Add State_Met as an argument to Init_CloudJ and Set_Clim_Profiles
- Now reference State_Met%MaxChemLev instead of State_Grid%MaxChemLev
- Cosmetic changes (indentation)

GeosCore/gc_environment_mod.F90
- Now pass State_Met as an argument to GC_Init_Extra
- Now reference State_Met%MaxChemLev instead of State_Grid%MaxChemLev

GeosCore/gc_grid_mod.F90
- Simplify logic where isGMAO and State_Grid%NativeNZ is computed

GeosCore/mercury_mod.F90
- Add State_Met argument to Init_Mercury and call to Init_Photolysis
- Now reference State_Met%MaxChemLev instead of State_Grid%MaxChemLev

GeosCore/photolysis_mod.F90
- Add State_Met argument to Init_Mercury and call to Init_CloudJ
- Now reference State_Met%MaxChemLev instead of State_Grid%MaxChemLev

GeosCore/sulfate_mod.F90
- Add State_Met argument to Init_Sulfate
- Now reference State_Met%MaxChemLev instead of State_Grid%MaxChemLev

GeosCore/aerosol_mod.F90
GeosCore/carbon_mod.F90
GeosCore/diagnostics_mod.F90
GeosCore/dust_mod.F90
GeosCore/global_br_mod.F90
GeosCore/hco_utilities_gc_mod.F90
GeosCore/ocean_mercury_mod.F90
GeosCore/seasalt_mod.F90
- Now reference State_Met%MaxChemLev instead of State_Grid%MaxChemLev

Interfaces/GCClassic/main.F90
Interfaces/GCHP/gchp_chunk_mod.F90
- Now pass State_Met to GC_Init_Extra and Init_Photolysis

Interfaces/GCHP/Chem_GridCompMod.F90
- Removed references to State_Grid%MaxStratLev and State_Grid%MaxTropLev

CHANGELOG.md
- Updated accordingly

Signed-off-by: Bob Yantosca <yantosca@seas.harvard.edu>

yantosca committed:
Bug fix: Now use State_Met%MaxChemLev in a couple more locations

GeosCore/aero_drydep.F90
- Change "State_Grid%MaxChemLev" -> "State_Met%MaxChemLev"

GeosCore/fjx_interface_mod.F90
- Change "State_Grid%MaxChemLev" -> "State_Met%MaxChemLev"
- Pass State_Met instead of State_Grid to "Set_Prof_FJX"
- Cosmetic changes (indentation)

GeosCore/gc_grid_mod.F90
- Fix typo "ModelE2.1" -> "MODELE2.1" in case statement where
  is_GMAO is set.  This needs to be in uppercase.

Signed-off-by: Bob Yantosca <yantosca@seas.harvard.edu>

yantosca commented:
We have rebased this PR atop e50dd60.

msulprizio commented on GeosCore/cldj_interface_mod.F90:
```diff
@@ -68,13 +69,14 @@ SUBROUTINE INIT_CLOUDJ( Input_Opt, State_Grid, State_Diag, State_Chm, RC )
     TYPE(OptInput), INTENT(IN)     :: Input_Opt   ! Input Options object
     TYPE(GrdState), INTENT(IN)     :: State_Grid  ! Grid State object
     TYPE(DgnState), INTENT(IN)     :: State_Diag  ! Diagnostics State object
+    TYPE(MetState), INTENT(IN)     :: State_Met   ! Meteorology State object
```
Why `State_Met` this added as input here? It doesn't seem to be used in this subroutine

yantosca replied:
State_Met is now removed from Init_CloudJ in 509eed8.

msulprizio commented on CHANGELOG.md:
```diff
@@ -22,6 +22,12 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
+- Added `State_Met` argument to routines `Init_CloudJ`, `Init_Photolysis`, `Set_Clim_Profiles`, `GC_Init_Extra`, `Init_Pressure`, `Init_Mercury`, `Init_Sulfate`
+- Added `State_Met%MaxChemLev` and `State_Met%MaxStratLev` integer fields
+- Added `Init_MaxChemLev` routine in `GeosUtil/pressure_mod.F90`, called from routine `Init_Pressure`
+- Added `State_Met` argument to routines `Init_CloudJ`, `Init_Photolysis`, `Set_Clim_Profiles`, `GC_Init_Extra`, `Init_Pressure`, `Init_Mercury`, `Init_Sulfate`, and `Set_Prof_FJX`
```
These three lines are repeated twice

yantosca replied:
Now fixed in 509eed8.

msulprizio commented on CHANGELOG.md:
```diff
@@ -48,6 +56,9 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - Removed `GeosUtil/grid_registry_mod.F90`.
 - Removed `OHconcAfterChem` from GCClassic and GCHP `HISTORY.rc.carbon` templates, as OH is fixed during the simulation
+- Removed `State_Grid%MaxChemLev`, `State_Grid%MaxStratLev`, and `State_Grid%MaxTropLev` fields
+- Removed `State_Grid%MaxChemLev`, `State_Grid%MaxStratLev`, and `State_Grid%MaxTropLev` fields
```
This line is the same as the previous line.

Also, I wonder if we can combine this comment on "removed `State_Grid$MaxChemLev` and `State_Grid%MaxStratLev`" and the comment above for "added `State_Met%MaxChemLev` and `State_Met%MaxStratLev` integer fields" to simply reflect that `MaxChemLev` and `MaxStratLev` were moved/renamed from `State_Grid` to `State_Met`.

yantosca replied:
Now fixed in 509eed8.

yantosca committed:
Removed State_Met from subroutine Init_CloudJ

GeosCore/cldj_interface_mod.F90
- Removed State_Met from the arg list, it's not used

GeosCore/photolysis_mod.F90
- Removed State_Met from the call to Init_CloudJ

CHANGELOG.md
- Updated accordingly
- Removed/combined duplicated text

Signed-off-by: Bob Yantosca <yantosca@seas.harvard.edu>

yantosca commented:
@msulprizio: The integration tests for GCC all pass but there are slight diffs in the ModelE2.1 test:

<img width="610" height="683" alt="gcap" src="https://github.com/user-attachments/assets/279a2733-80f8-48fe-b59c-0c160c47de8f" />

There are also diffs in the GCHP fullchem tests, but this looks more like numerical noise.
<img width="633" height="678" alt="gchp" src="https://github.com/user-attachments/assets/141917eb-c334-4d78-bfaa-bd92228d6dca" />

I would hold off on merging until we can figure out what is happening.

yantosca committed:
Use hardwired MaxChemLev, MaxStratLev for GCAP/ModelE2.1

GeosUtil/pressure_mod.F90
- Added a CASE statement so that State_Met%MaxChemLev and
  State_Met%MaxStratLev will be hardwired for the GCAP/ModelE2.1
  102, 74, and 40 level grids.  For the GMAO grids these will
  be computed as the  1hPa level.

This was necessary in order to obtain zero-diff results when using
GCAP ModelE2.1 met fields when comparing to previous model versions.

Signed-off-by: Bob Yantosca <yantosca@seas.harvard.edu>

yantosca commented:
@msulprizio: I've restored the original definitions of MaxChemLev and MaxStratLev for the GCAP ModelE2.1 grids, while computing these as the 1 hPa level in commit https://github.com/geoschem/geos-chem/pull/3089/commits/51762a4fdaa370245f064ff0afd88377f8e8ae1b.  This now results in 100% bitwise identical results w/r/t older model versions when using GCAP ModelE2.1 meteorology.

<img width="610" height="698" alt="gcap2" src="https://github.com/user-attachments/assets/ecf4a59a-e847-48e4-8857-93091212c813" />

yantosca committed:
Removed leftover State_Grid%MaxChemLev, State_Grid%MaxStratLev instances

Headers/state_grid_mod.F90
- Removed MaxChemLev, MaxStratLev, MaxTropLev from TYPE(GrdState)

Interfaces/GCHP/Chem_GridCompMod.F90
- Changed: State_Grid%MaxChemLev  -> State_Met%MaxGridLev
- Changed: State_Grid%MaxStratLev -> State_Met%MaxStratLev

Interfaces/GCHP/gchp_chunk_mod.F90
- Removed references to State_Grid%MaxChemLev, State_Grid%MaxStratLev

Signed-off-by: Bob Yantosca <yantosca@seas.harvard.edu>

yantosca commented:
Found & fixed some GCHP bugs, rerunning integration tests

yantosca commented:
I fixed the differences in GCHP in commit 7ff7bb5.  There were some leftover instances of `State_Grid%MaxChemLev` and `State_Grid%MaxStratLev` in `Chem_GridCompMod.F90` and `gchp_chunk_mod.F90` that were masking the correct values of `State_Met%MaxChemLev` and `State_Met%MaxStratLev`.  

After having fixed this issue, I re-ran integration tests on this branch and the differences in GCHP disappeared.

All GEOS-Chem Classic integration tests passed:
```console
==============================================================================
GEOS-Chem Classic: Execution Test Results

CodeDir   : 81a03ce GEOS-Chem update: Merge PR #3205 (Fix bugs in HEMCO_Diagn.rc templates)
GEOS-Chem : 7ff7bb514 Removed leftover State_Grid%MaxChemLev, State_Grid%MaxStratLev instances
HEMCO     : 9fbb78e Merge PR #355 (Update to latest GitHub Actions versions)
Cloud-J   : 16d18b0 Merge PR #34 (Fixed issues in format statements for verbose output)
HETP      : 2a99b24 Merge pull request #2 from geoschem/bugfix/initialize_local_variables

Using 24 OpenMP threads
Number of execution tests: 38

Submitted as SLURM job: 3801929
==============================================================================

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%  All execution tests passed!  %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```
All GEOS-Chem Classic tests were zero-diff w/r/t `gcc.7.285af20`,except

- gc_4x5_merra2_fullchem_APM
  - Parallel error (known issue)


All GCHP integration tests passed:
```console
==============================================================================
GCHP: Execution Test Results

CodeDir       : 95b418f GEOS-Chem update: Merge PR #3205 (Fix bugs in HEMCO_Diagn.rc templates)
MAPL          : 80d66d0a Merge PR #41 (ExtData2G in GCHP Transport Tracers simulation)
GMAO_Shared   : 4ddb3ec Merge pull request #2 from geoschem/feature/mapl-upgrade
ESMA_cmake    : 01bbf2f Merge pull request #9 from geoschem/feature/turn_off_f2py_builds_in_gchp
FMS           : 259759d1 Merge pull request #3 from geoschem/feature/update_gmao_libs
FVdycoreCubed : 348620f Merge PR #11 (Scale tracers by pressure ratio in first timestep)
geos-chem     : 7ff7bb514 Removed leftover State_Grid%MaxChemLev, State_Grid%MaxStratLev instances
HEMCO         : 9fbb78e Merge PR #355 (Update to latest GitHub Actions versions)
Cloud-J       : 16d18b0 Merge PR #34 (Fixed issues in format statements for verbose output)
HETP          : 2a99b24 Merge pull request #2 from geoschem/bugfix/initialize_local_variables
GFE           : bac529f Merge pull request #40 from Goddard-Fortran-Ecosystem/feature/mathomp4/update-gftl-1.15.2

Number of execution tests: 12

Submitted as SLURM job: 3801733
==============================================================================

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%  All execution tests passed!  %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```
All GCHP tests were zero-diff w/r/t `gchp.7.63b9257`.

msulprizio commented on GeosCore/cldj_interface_mod.F90:
```diff
@@ -68,7 +69,7 @@ SUBROUTINE INIT_CLOUDJ( Input_Opt, State_Grid, State_Diag, State_Chm, RC )
     USE State_Grid_Mod, ONLY : GrdState
     USE State_Diag_Mod, ONLY : DgnState
-
+    USE State_Met_Mod,  ONLY : MetState
```
This line can be removed

yantosca replied:
Now resolved in e1a615ab

yantosca committed:
Removed leftover USE statement from Init_CloudJ routine

GeosCore/cldj_interface_mod.F90
- Removed the leftover "USE State_Met_Mod" statement.  This is not
  needed in this routine.

Signed-off-by: Bob Yantosca <yantosca@seas.harvard.edu>

yantosca commented:
All GEOS-Chem Classic integration tests passed:
```console
==============================================================================
GEOS-Chem Classic: Execution Test Results

CodeDir   : 0917c93 GEOS-Chem update: Merge PR #3089 (Remove hardwired MaxChemLev)
GEOS-Chem : 004d762e2 Merge PR #3089 (Remove hardwired MaxChemLev and MaxStratLev)
HEMCO     : 9fbb78e Merge PR #355 (Update to latest GitHub Actions versions)
Cloud-J   : 16d18b0 Merge PR #34 (Fixed issues in format statements for verbose output)
HETP      : 2a99b24 Merge pull request #2 from geoschem/bugfix/initialize_local_variables

Using 24 OpenMP threads
Number of execution tests: 38

Submitted as SLURM job: 4018903
==============================================================================

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%  All execution tests passed!  %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```
All GEOS-Chem Classic tests were zero-diff w/r/t `gcc.7.285af20`, except

- gc_4x5_merra2_fullchem_APM
  - Parallel error (known issue)

All GCHP integration tests passed:
```console
==============================================================================
GCHP: Execution Test Results

CodeDir       : 84a1244 GEOS-Chem update: Merge PR #3089 (Remove hardwired MaxChemLev)
MAPL          : 80d66d0a Merge PR #41 (ExtData2G in GCHP Transport Tracers simulation)
GMAO_Shared   : 4ddb3ec Merge pull request #2 from geoschem/feature/mapl-upgrade
ESMA_cmake    : 01bbf2f Merge pull request #9 from geoschem/feature/turn_off_f2py_builds_in_gchp
FMS           : 259759d1 Merge pull request #3 from geoschem/feature/update_gmao_libs
FVdycoreCubed : 348620f Merge PR #11 (Scale tracers by pressure ratio in first timestep)
geos-chem     : 004d762e2 Merge PR #3089 (Remove hardwired MaxChemLev and MaxStratLev)
HEMCO         : 9fbb78e Merge PR #355 (Update to latest GitHub Actions versions)
Cloud-J       : 16d18b0 Merge PR #34 (Fixed issues in format statements for verbose output)
HETP          : 2a99b24 Merge pull request #2 from geoschem/bugfix/initialize_local_variables
GFE           : bac529f Merge pull request #40 from Goddard-Fortran-Ecosystem/feature/mathomp4/update-gftl-1.15.2

Number of execution tests: 12

Submitted as SLURM job: 4019281
==============================================================================

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%  All execution tests passed!  %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```
All GCHP tests were zero-diff w/r/t `gchp.7.63b9257`, except:

- gchp_merra2_fullchem_alldiags
- gchp_merra2_fullchem_benchmark
  - The update in PR #3205 restored the InvDustL23M diagnostic output to the `GEOSChem.Emissions.*.nc4` files.  This causes the emissions diagnostics to differ.  However, the restart files are identical and this means that there are no changes to the scientific results.
